_source/_posts/2020-04-20-android-authentication.adoc

```diff
@@ -0,0 +1,508 @@
+   private val promptInfo: PromptInfo
+
+   init {
+       prompt = BiometricPrompt(fragmentActivity, Executors.newSingleThreadExecutor(), mCallback)
+       promptInfo = PromptInfo.Builder()
+           .setTitle("Biometric authentication succeeded")
+           .setDeviceCredentialAllowed(true)
+           .setConfirmationRequired(true)
+           .build()
+   }
+
+   fun show() {
+       prompt.authenticate(promptInfo)
+   }
+}
+----
+
+Simply trigger this so that when a user who is not biometrically authenticated tries to launch `MainActivity`, they are only allowed through if they biometrically authenticate in their device.
+Also, remember to use `GuardedEncryptionManager` to store Okta OIDC data:
+
+[source,kotlin]
+----
+import com.okta.oidc.storage.security.DefaultEncryptionManager
+import com.okta.oidc.storage.security.EncryptionManager
+import com.okta.oidc.storage.security.GuardedEncryptionManager
+
+private var currentEncryptionManager: EncryptionManager? = null
+private var keyguardEncryptionManager: GuardedEncryptionManager? = null
+
+override fun onResume() {
+   super.onResume()
+
+   // Check whether the user has already authenticated using the device's authentication method. If it hasn't, ask them to do it
+   if (currentEncryptionManager?.isUserAuthenticatedOnDevice?.not() == true) {
+       showKeyguard()
+   }
+}
+
+private fun showKeyguard() {
+   Biometric(
+       fragmentActivity = this,
+       onSuccessListener = {
+           Log.d("MainActivity", "Biometric authentication succeeded")
+           Toast.makeText(this, "Biometric authentication succeeded", Toast.LENGTH_SHORT).show()
+       },
+       onCancelListener = {
+           runOnUiThread {
+               Log.d("MainActivity", "Biometric authentication cancelled")
+               Toast.makeText(this, "Biometric authentication cancelled", Toast.LENGTH_SHORT).show()
+               finish()
+           }
+       },
+       onErrorListener = { code, message ->
+           runOnUiThread {
+               Log.d("MainActivity", "Biometric authentication failed")
+               Toast.makeText(this, "Biometric authentication failed", Toast.LENGTH_SHORT).show()
+               finish()
+           }
+       }).show()
+}
+
+private fun setupWebAuth() {
+   keyguardEncryptionManager = GuardedEncryptionManager(this, Int.MAX_VALUE)
+
+   webAuth = WebAuthBuilder()
+       .withConfig(oidcConfig)
+       .withContext(applicationContext)
+       .withCallbackExecutor(null)
+       .withEncryptionManager(keyguardEncryptionManager)
+       .create()
+   sessionClient = webAuth.sessionClient
+}
+----
+
+Now, whenever a user opens a new instance of your app, they will be able to access it using their biometric information. In the below example, a Google Pixel phone, the authentication is a  fingerprint:
+
+image::{% asset_path 'blog/android-authentication/android-fingerprint-prompt.png' %}[alt=Android fingerprint prompt,width=400,align=center]
+
+== Summary
```

Comment:
Can this be made better for SEO? Most of the time, we combine the summary and the learn more.